### PR TITLE
Implement Application Branding Resolver

### PR DESCRIPTION
**Pull request by dhaura**
> Refactor Organization Branding Resolver accroding to [1].

> 1. Check whether the shared application has configured app-level branding.
> 2. If step 1 is not found, check the shared app residing sub-organization's org-level branding.
> 3. If step 2 is not found, check whether the shared app in the immediate parent organization has app-level branding.
> 4. If step 3 is not found, get the org-level branding from the immediate parent organization.
> 5. If step 4 is not found, repeat 3 and 4 throughout the organization hierarchy until branding is found at some level.
> 6. If branding is not found after step 5, default branding is applied.

abrp-2024-06-14-1144

root-inheritance

parent-inheritance

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

## Related Issue 
- https://github.com/wso2/product-is/issues/20616
- https://github.com/wso2/product-is/issues/16846

## Related PRs
- https://github.com/wso2-extensions/identity-organization-management/pull/383

 

---

## References
[1] - https://github.com/wso2/product-is/issues/16846

### CODE REVIEW

File: components/org.wso2.carbon.identity.branding.preference.management.core/src/main/java/org/wso2/carbon/identity/branding/preference/management/core/constant/BrandingPreferenceMgtConstants.java [hunk 1]
```diff
@@ -54,7 +54,7 @@ public enum ErrorMessages {
         ERROR_CODE_INVALID_BRANDING_PREFERENCE("BRANDINGM_00001",
                 "Invalid Branding Preference configurations for tenant: %s."),
         ERROR_CODE_BRANDING_PREFERENCE_NOT_EXISTS("BRANDINGM_00002",
-                "Branding preferences are not configured for tenant: %s."),
+                "Branding preferences are not configured for name: %s, type: %s, tenant: %s."),
```
Comment (dhaura) [inline components/org.wso2.carbon.identity.branding.preference.management.core/src/main/java/org/wso2/carbon/identity/branding/preference/management/core/constant/BrandingPreferenceMgtConstants.java:57]: Try to go for a new constant

File: components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java [hunk 2]
```diff
@@ -117,10 +132,23 @@ public BrandingPreference resolveBranding(String type, String name, String local
             }
         }
 
-        /* Tenant domain will always be carbon.super for SaaS apps (ex. myaccount). Hence need to resolve
+        if (APPLICATION_TYPE.equals(type)) {
+            return resolveApplicationBranding(name, locale, organizationManager, organizationId, currentTenantDomain);
+        } else if (ORGANIZATION_TYPE.equals(type)) {
+            return resolveOrganizationBranding(name, locale, organizationManager, organizationId, currentTenantDomain);
+        } else {
+            throw handleClientException(ERROR_CODE_INVALID_BRANDING_PREFERENCE_TYPE, type, getTenantDomain());
```
Comment (dhaura) [inline components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java:140]: getTenantDomain() can be replaced with currentTenantDomain

File: components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java [hunk 3]
```diff
@@ -203,15 +227,144 @@ Get ancestor organization ids (including itself) of a given organization. The li
 
             // No branding found. Adding the same tenant domain to cache to avoid the resolving in the next run.
             addToCache(organizationId, currentTenantDomain, currentTenantDomain);
-            throw handleClientException(ERROR_CODE_BRANDING_PREFERENCE_NOT_EXISTS, getTenantDomain());
+            throw handleClientException(ERROR_CODE_BRANDING_PREFERENCE_NOT_EXISTS,
+                    name, ORGANIZATION_TYPE, getTenantDomain());
```
Comment (dhaura) [inline components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java:231]: getTenantDomain() can be replaced with currentTenantDomain

Comment (dhaura) [reply components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java:231]: Check other places

File: components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java [hunk 4]
```diff
@@ -203,15 +227,144 @@ Get ancestor organization ids (including itself) of a given organization. The li
...
+                    addAppBrandingToCache(appId, currentTenantDomain, null, parentTenantDomain, ORGANIZATION_TYPE);
+                    return brandingPreference.get();
+                }
+
+                /*
+                    Get ancestor organization ids (including itself) of a given organization. The list is sorted
+                    from given organization id to the root organization id.
+                */
+                List<String> ancestorOrganizationIds = organizationManager.getAncestorOrganizationIds(parentOrgId);
+                if (!ancestorOrganizationIds.isEmpty() && ancestorOrganizationIds.size() > 1) {
```
Comment (dhaura) [inline components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java:343]: Can getAncestorOrganizationIds return null?

File: components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java [hunk 5]
```diff
@@ -203,15 +227,144 @@ Get ancestor organization ids (including itself) of a given organization. The li
...
+                    // Go to the parent organization again.
+                    parentAppId = orgApplicationManager.getParentAppId(parentAppId, parentOrgId);
+                    parentOrgId = ancestorOrganizationIds.get(1);;
```
Comment (dhaura) [inline components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java:346]: Remove `;`

File: components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java [hunk 6]
```diff
@@ -203,15 +227,144 @@ Get ancestor organization ids (including itself) of a given organization. The li
...
+            OrgApplicationManager orgApplicationManager =
+                    BrandingResolverComponentDataHolder.getInstance().getOrgApplicationManager();
+            String parentAppId = orgApplicationManager.getParentAppId(appId, orgId);
+            String parentTenantDomain = organizationManager.resolveTenantDomain(parentOrgId);
+            int parentDepthInHierarchy = organizationManager.getOrganizationDepthInHierarchy(parentOrgId);
+
+            // Get the minimum hierarchy depth that needs to be reached to resolve branding preference
+            int minHierarchyDepth = Utils.getSubOrgStartLevel() - 1;
+
+            while (parentDepthInHierarchy >= minHierarchyDepth) {
```
Comment (dhaura) [inline components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java:311]: Refactor according to https://github.com/wso2/product-is/issues/16846

File: components/org.wso2.carbon.identity.branding.preference.management.core/src/main/java/org/wso2/carbon/identity/branding/preference/management/core/UIBrandingPreferenceResolver.java [hunk 7]
```diff
@@ -39,13 +39,27 @@ public interface UIBrandingPreferenceResolver {
...
+    /**
+     * This method is used to clear the branding preference resolver caches, down
+     * the organization hierarchy.
+     *
+     * @param type                Branding preference type.
+     * @param name                Branding preference name.
+     * @param currentTenantDomain Tenant domain where the cache needs to be cleared.
+     * @throws BrandingPreferenceMgtException if any error occurred.
+     */
+    void clearBrandingResolverCacheHierarchy(String type, String name, String currentTenantDomain)
```
Comment (dhaura) [inline components/org.wso2.carbon.identity.branding.preference.management.core/src/main/java/org/wso2/carbon/identity/branding/preference/management/core/UIBrandingPreferenceResolver.java:51]: Make it default.

File: components/org.wso2.carbon.identity.branding.preference.management.core/src/main/java/org/wso2/carbon/identity/branding/preference/management/core/constant/BrandingPreferenceMgtConstants.java (hunk 1 above)
Comment (dhaura) [reply components/org.wso2.carbon.identity.branding.preference.management.core/src/main/java/org/wso2/carbon/identity/branding/preference/management/core/constant/BrandingPreferenceMgtConstants.java]: Addressed with https://github.com/wso2-extensions/identity-branding-preference-management/pull/40/commits/05fa6da03cb482ecbd514d7030744011bad7a98e

File: components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java (hunk 2 above)
Comment (dhaura) [reply components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java]: Addressed with https://github.com/wso2-extensions/identity-branding-preference-management/pull/40/commits/05fa6da03cb482ecbd514d7030744011bad7a98e

File: components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java (hunk 3 above)
Comment (dhaura) [reply components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java]: Addressed with https://github.com/wso2-extensions/identity-branding-preference-management/pull/40/commits/05fa6da03cb482ecbd514d7030744011bad7a98e

File: components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java (hunk 5 above)
Comment (dhaura) [reply components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java]: Addressed with https://github.com/wso2-extensions/identity-branding-preference-management/pull/40/commits/05fa6da03cb482ecbd514d7030744011bad7a98e

File: components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java (hunk 6 above)
Comment (dhaura) [reply components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java]: Addressed with https://github.com/wso2-extensions/identity-branding-preference-management/pull/40/commits/05fa6da03cb482ecbd514d7030744011bad7a98e

File: components/org.wso2.carbon.identity.branding.preference.management.core/src/main/java/org/wso2/carbon/identity/branding/preference/management/core/UIBrandingPreferenceResolver.java (hunk 7 above)
Comment (dhaura) [reply components/org.wso2.carbon.identity.branding.preference.management.core/src/main/java/org/wso2/carbon/identity/branding/preference/management/core/UIBrandingPreferenceResolver.java]: Addressed with https://github.com/wso2-extensions/identity-branding-preference-management/pull/40/commits/05fa6da03cb482ecbd514d7030744011bad7a98e

File: components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java [hunk 8]
```diff
@@ -82,20 +94,25 @@ public class UIBrandingPreferenceResolverImpl implements UIBrandingPreferenceRes
...
     /**
      * UI branding preference resolver implementation constructor.
      *
-     * @param brandedOrgCache Cache instance.
+     * @param brandedOrgCache        Cache instance for branded org.
      * @param textCustomizedOrgCache Cache instance for custom text.
+     * @param brandedAppCache        Cache instance for branded app.
      */
-    public UIBrandingPreferenceResolverImpl(BrandedOrgCache brandedOrgCache,
+    public UIBrandingPreferenceResolverImpl(BrandedOrgCache brandedOrgCache, BrandedAppCache brandedAppCache,
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java:111]: This is a change to a public constructor.
Introduce a new one

File: components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java [hunk 9]
```diff
@@ -82,20 +94,25 @@ public class UIBrandingPreferenceResolverImpl implements UIBrandingPreferenceRes
...
+    private final BrandedAppCache brandedAppCache;
     private final TextCustomizedOrgCache textCustomizedOrgCache;
 
     /**
      * UI branding preference resolver implementation constructor.
      *
-     * @param brandedOrgCache Cache instance.
+     * @param brandedOrgCache        Cache instance for branded org.
      * @param textCustomizedOrgCache Cache instance for custom text.
+     * @param brandedAppCache        Cache instance for branded app.
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java:109]: order is different

File: components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java [hunk 10]
```diff
@@ -117,10 +134,23 @@ public BrandingPreference resolveBranding(String type, String name, String local
             }
         }
 
-        /* Tenant domain will always be carbon.super for SaaS apps (ex. myaccount). Hence need to resolve
+        if (APPLICATION_TYPE.equals(type)) {
+            return resolveApplicationBranding(name, locale, organizationManager, organizationId, currentTenantDomain);
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java:138]: `organizationManager` is not a matching param for the private method.
Seems you have passed it to use the same instence service and reduce the code.
but for the clean method signature better to remove it and get organizationManager inside the private methods.

File: components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java [hunk 11]
```diff
@@ -117,10 +134,23 @@ public BrandingPreference resolveBranding(String type, String name, String local
...
+        } else if (ORGANIZATION_TYPE.equals(type)) {
+            return resolveOrganizationBranding(name, locale, organizationManager, organizationId, currentTenantDomain);
+        } else {
+            throw handleClientException(ERROR_CODE_INVALID_BRANDING_PREFERENCE_TYPE, type, currentTenantDomain);
+        }
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java:143]:
```suggestion
        }
        throw handleClientException(ERROR_CODE_INVALID_BRANDING_PREFERENCE_TYPE, type, currentTenantDomain);
```

File: components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java (hunk 8 above)
Comment (dhaura) [reply components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java:124]: Addressed with https://github.com/wso2-extensions/identity-branding-preference-management/pull/40/commits/a9322948eef77b783bd87838edceb408a676e7f4

File: components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java (hunk 9 above)
Comment (dhaura) [reply components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java]: Addressed with https://github.com/wso2-extensions/identity-branding-preference-management/pull/40/commits/a9322948eef77b783bd87838edceb408a676e7f4

File: components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java (hunk 10 above)
Comment (dhaura) [reply components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java]: Addressed with https://github.com/wso2-extensions/identity-branding-preference-management/pull/40/commits/a9322948eef77b783bd87838edceb408a676e7f4

File: components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java (hunk 11 above)
Comment (dhaura) [reply components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java]: Addressed with https://github.com/wso2-extensions/identity-branding-preference-management/pull/40/commits/a9322948eef77b783bd87838edceb408a676e7f4

**Review by AnuradhaSK — APPROVED** [review]

**Pull request merged**